src/libraries/Microsoft.Extensions.DependencyInjection/src/DependencyInjectionEventSource.cs

```diff
@@ -29,6 +30,8 @@ private DependencyInjectionEventSource() : base(EventSourceSettings.EtwSelfDescr
         // - A stop event's event id must be next one after its start event.
         // - Avoid renaming methods or parameters marked with EventAttribute. EventSource uses these to form the event object.
 
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
+            Justification = "Parameters to this method are primitive and are trimmer safe.")]
```

Comment:
Can you help me understand the linker warning? I
e.g. why is this method an issue but ExpressionTreeGenerated below is not?

Reply:
The linker will throw this warning in a method if any of the methods it calls has a RequiresUnreferencedCode attribute on it. The WriteEvent overload that this method is calling has the attribute:
https://github.com/dotnet/runtime/blob/493ba8d90f96adc6c3a218c7fad3428ebac247ec/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs#L1404-L1410

while the one that ExpressionTreeGenerated does not. In fact, that one was getting a warning itself, but it was deemed safe so it suppressed the warning, meaning callers to that method would not get a warning themselves:

https://github.com/dotnet/runtime/blob/493ba8d90f96adc6c3a218c7fad3428ebac247ec/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs#L970-L977

Reply:
Got it.  There is no built-in WriteEvent overload that takes string/string/int/int, so it's binding to the object[] args overload, which then is annotated as RequiresUnreferencedCode because it may need to reflect over arbitrary objects in that args list.  In contrast, the event I cited below passes just string/int, which binds to a built-in string/int-based overload, and because those argument types are fully understood, the warning there on the helper is suppressed.  Thanks.

Reply:
Your summary is much better than my lame attempt at explaining 😆 but yeah, you are exactly right 😄